zkevm-circuits/src/evm_circuit/execution/returndatacopy.rs

```diff
@@ -0,0 +1,359 @@
+
+#[derive(Clone, Debug)]
+pub(crate) struct ReturnDataCopyGadget<F> {
+    same_context: SameContextGadget<F>,
+    /// Holds the memory address for return data from where we read.
+    return_data_offset: Cell<F>, //MemoryAddress<F>,
+    /// Holds the size of the return data.
+    return_data_size: Cell<F>, //RandomLinearCombination<F, N_BYTES_MEMORY_ADDRESS>,
+    /// The data is copied to memory. To verify this
+    /// copy operation we need the MemoryAddressGadget.
+    dst_memory_addr: MemoryAddressGadget<F>,
+    /// Holds the memory address for the offset in return data from where we
+    /// read.
+    data_offset: MemoryAddress<F>,
+    /// Opcode RETURNDATACOPY has a dynamic gas cost:
+    /// gas_code = static_gas * minimum_word_size + memory_expansion_cost
+    memory_expansion: MemoryExpansionGadget<F, 1, N_BYTES_MEMORY_WORD_SIZE>,
+    /// Opcode RETURNDATAECOPY needs to copy data into memory. We account for
+    /// the copying costs using the memory copier gas gadget.
+    memory_copier_gas: MemoryCopierGasGadget<F, { GasCost::COPY }>,
+    /// RW inverse counter from the copy table at the start of related copy
+    /// steps.
+    copy_rwc_inc: Cell<F>,
+    /// Out of bound check circuit.
+    in_bound_check: LtGadget<F, N_BYTES_MEMORY_WORD_SIZE>,
+}
+
+impl<F: Field> ExecutionGadget<F> for ReturnDataCopyGadget<F> {
+    const NAME: &'static str = "RETURNDATACOPY";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::RETURNDATACOPY;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        // let dest_offset = cb.query_rlc();
+        let dest_offset = cb.query_cell();
+        let data_offset = cb.query_rlc();
+        let size = cb.query_rlc();
+
+        // 1. Pop dest_offset, offset, length from stack
+        cb.stack_pop(dest_offset.expr());
+        cb.stack_pop(data_offset.expr());
+        cb.stack_pop(size.expr());
+
+        // 2. Add lookup constraint in the call context for the returndatacopy field.
+        let return_data_offset = cb.query_cell();
+        let return_data_size = cb.query_cell();
+        cb.call_context_lookup(
+            false.expr(),
+            None,
+            CallContextFieldTag::LastCalleeReturnDataOffset,
+            return_data_offset.expr(),
+        );
+        cb.call_context_lookup(
+            false.expr(),
+            None,
+            CallContextFieldTag::LastCalleeReturnDataLength,
+            return_data_size.expr(),
+        );
+
+        // 3. contraints for copy: copy overflow check
+        // i.e., offset + size <= return_data_size
+        let in_bound_check = LtGadget::construct(
+            cb,
+            from_bytes::expr(&data_offset.cells) + size.expr(),
+            return_data_size.expr() + 1.expr(),
+        );
+        cb.require_equal(
+            "offset + size < return_data_size + 1",
+            in_bound_check.expr(),
+            1.expr(),
+        );
```

Comment:
Can be simplified I think as 
```
RangeCheckGadget::construct(
  cb, 
  return_data_size.expr() - (from_bytes::expr(&data_offset.cells) + from_bytes::expr(&size.cells)
);
```

Also `size.expr()` is used in your code, but it should be `from_bytes::expr(&size.cells)` no?

Reply:
Oh, should be from_bytes::expr(). fixed. hmm, It weird to see all test passed.

Here I actually have a question, there are 2 ways to build a value (in single cell or a bunch of cells).
1. a = cb.query_cell(), later, a.expr() can be used if some other constraints refer to a.
2. b = cb.query_rlc(), then, from_bytes::expr(b.cells) is used.
At the beginning, I use rlc (or cell depends on data type), and after some debugging, I changed some to the other way, but it seems both are ok, I just wonder is there a standard to choose between cell & rlc??

Reply:
If the value can fit in a single field element you generally just use `query_cell()`. If the value cannot fit in a single field element (like an EVM word that has size 256 bits) you generally split up the value in bytes (for now) and do the calculations in parts (like add, the comparisons, etc...). In some cases the value comes from somewhere else, like in this example where the `size` comes from the stack. All stack values are encoded and stored using rlc, and so you need the rlc encoding just be able to read the value from the stack (and stack values are 256 bit words in general). So in this case you could argue you could store the size in a single field elements (because otherwise it would either revert or go out of gas), but you have to do the rlc encoding regardless (which requires breaking up the values in bytes) because it's also directly used to pop it from the stack, so `query_rlc()` is the way to go.

Reply:
Yes, that's also my understanding when I read the spec, it looks that the data comes from stack has the RLC encoding. But the code references do not comply with this.... That's why I changes the query way several times.
You can see 
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/f3320827b90afde4e48589527294d1fd2ca77a72/zkevm-circuits/src/evm_circuit/execution/codecopy.rs#L53-L60
and
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/f3320827b90afde4e48589527294d1fd2ca77a72/zkevm-circuits/src/evm_circuit/execution/calldatacopy.rs#L53-L60

The stack pops memory_offset out, which is a cell generated by query_cell. But in spec, they are in RLC format as expected, see  

https://github.com/privacy-scaling-explorations/zkevm-specs/blob/e3bbbc6070579069d868d4ffc003d0c627d81e3f/tests/evm/test_codecopy.py#L67-L69

If the memory_offset is not a special case, the code is wrong, otherwise, the spec is wrong.
BTW: I follow the code reference, so the spec is also inconsistent with the code in this line. 

what a mess :(

Reply:
I think it still makes sense, just a bit confusing. `MemoryAddressGadget::construct` does the rlc decomposition for the offset, so the input can just be the rlc encoded value which can be stored a single cell. So in cases where you don't actually need to have access to the actual value (like in pop where just the rlc encoded value is enough) or in cases where you only need to check if the value is `0` (because the rlc encoded value of `0` is still `0`) you don't need to have the byte decomposition because you don't need it.

So I think this is just a bit of an optimization for certain opcodes (or an unfortunate API decision for the gadget) which the spec doesn't do.

Reply:
That's ok, is the optimization in stack pop, whichi means stack pop rlc value has more cost than pop single cell value? Otherwise I can not see many benefits as MemoryAddressGadget queried another rlc cells and make an extra equal constraint between cell passed in & local rlc cells?

Reply:
> It weird to see all test passed since size.expr() was wrong.

Just figured out that the test passed is because the size param in all tests are small (in [0, 255]), so rlc == size. The rlc value differs if the size has >1 bytes. Added few more test cases to make rlc really take effect.